File: 0210-course-schedule-ii/0210-course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        
        d = {i: [] for i in range(numCourses)}
        vis = {i: 0 for i in range(numCourses)}
        
        for crs, pre in prerequisites:
            d[crs].append(pre)
            
        res = []
        
        def dfs(i):
            if vis[i]==1:
                return True
            
            if vis[i]==0:
                vis[i]=1
                for x in d[i]:
                    if dfs(x):
                        return True
                    
                vis[i] = 2
                
                res.append(i)
                
            return False
        
        for i in range(numCourses):
            if dfs(i):
                return []
            
        return res

File: 0210-course-schedule-ii/0210-course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int,vector<int>> adj;
        for(auto e:prerequisites){
            adj[e[0]].push_back(e[1]);
        }
        
        vector<int> res,vis(numCourses,0);
        for(int i=0;i<numCourses;i++)
            if(!dfs(adj, i, vis, res))
                return {};
        
        return res;
    }
    
    bool dfs(unordered_map<int,vector<int>> &adj, int s, vector<int> &vis, vector<int> &res){
        if(vis[s]==1){
            return false;
        }
        else if(vis[s]==0){
            vis[s]=1;
            for(auto nei:adj[s]){
                if(!dfs(adj,nei,vis,res))
                    return false;
            }
            res.push_back(s);
            vis[s]=2;
        }
        return true;
    }
};